File: server/lib/fazoura/game/pack.ex
defmodule Fazoura.Game.Pack do
  @moduledoc """
  Immutable snapshot of the questions a running room plays.

  Built from a stored quiz by `Fazoura.Quizzes.to_pack/1` (or `from_map/1` in tests), and
  from several of them by `merge/1` — a round is played from one pool drawn from every
  quiz the host selected (PROTOCOL.md §6.4). Rooms copy it at creation, so later edits can
  never affect a running game.
  """

  defmodule Question do
    @moduledoc false

    @enforce_keys [:id, :type, :prompt, :accepted_answers, :time_limit_ms]
    defstruct [
      :id,
      :type,
      :prompt,
      :accepted_answers,
      :time_limit_ms,
      # The stored quiz this question was snapshotted from, so a player who
      # sees something in a game can report it (QUIZ_FORMAT.md §5.9) without
      # anybody having to broadcast quiz ids — an id in a `state` would also be
      # a cheat button, since `GET /api/quizzes/:id/download` hands out the
      # accepted answers. `nil` for a private quiz sent inline: there is nothing
      # published to report.
      quiz_id: nil,
      image_url: nil,
      difficulty: "easy"
    ]

    @type t :: %__MODULE__{
            id: String.t(),
            type: String.t(),
            prompt: String.t(),
            accepted_answers: [String.t()],
            time_limit_ms: pos_integer(),
            quiz_id: String.t() | nil,
            image_url: String.t() | nil,
            difficulty: String.t()
          }
  end

  @enforce_keys [:titles, :questions]
  defstruct [
    :questions,
    # Titles of the selected quizzes, in the order the host chose them. Empty
    # means nothing is selected yet (PROTOCOL.md §5.1, §6.4).
    titles: [],
    # Suggested room settings, from the first quiz selected (QUIZ_FORMAT.md §2.1).
    default_time_limit_ms: nil,
    default_difficulty_multiplier: false
  ]

  @type t :: %__MODULE__{
          titles: [String.t()],
          questions: [Question.t()],
          default_time_limit_ms: pos_integer() | nil,
          default_difficulty_multiplier: boolean()
        }

  @default_time_limit_ms 30_000

  @doc """
  One pool from the quizzes the host selected, in the order they selected them
  (PROTOCOL.md §6.4).

  Question ids are rewritten to stay unique across the pool: two quizzes may each call a
  question `q1`, and clients use the id to tell one question from the next. Ids are opaque
  (§1), so rewriting them is ours to do. Lobby defaults come from the first quiz, because a
  pool has none of its own.
  """
  @spec merge([t()]) :: t()
  def merge([pack]), do: pack

  def merge([first | _] = packs) do
    %__MODULE__{
      titles: Enum.flat_map(packs, & &1.titles),
      default_time_limit_ms: first.default_time_limit_ms,
      default_difficulty_multiplier: first.default_difficulty_multiplier,
      questions:
        packs
        |> Enum.with_index()
        |> Enum.flat_map(fn {pack, index} ->
          Enum.map(pack.questions, &%{&1 | id: "#{index}-#{&1.id}"})
        end)
    }
  end

  @spec from_map(map()) :: t()
  def from_map(%{"title" => title, "questions" => questions}) do
    %__MODULE__{
      titles: [title],
      questions: Enum.map(questions, &question_from_map/1)
    }
  end

  defp question_from_map(map) do
    %Question{
      id: Map.fetch!(map, "id"),
      type: Map.get(map, "type", "text"),
      prompt: Map.fetch!(map, "prompt"),
      accepted_answers: Map.fetch!(map, "accepted_answers"),
      time_limit_ms: Map.get(map, "time_limit_ms", @default_time_limit_ms),
      image_url: Map.get(map, "image_url"),
      difficulty: difficulty!(Map.get(map, "difficulty", "easy"))
    }
  end

  @difficulties ~w(easy medium hard)

  defp difficulty!(value) when value in @difficulties, do: value

  defp difficulty!(value),
    do: raise(ArgumentError, "unknown difficulty #{inspect(value)}; use easy, medium or hard")
end

File: app/lib/core/game/pack.dart
/// Immutable snapshot of the questions a LAN-hosted room plays.
///
/// Dart port of `server/lib/fazoura/game/pack.ex`. A round is played from one
/// pool drawn from every quiz the host selected ([Pack.merge], PROTOCOL.md
/// §6.4). The room copies it when it is created, so editing a quiz mid-game
/// can never affect a running room (AGENTS.md §4).
library;

import '../models/quiz.dart';

const _defaultTimeLimitMs = 30000;
const _difficulties = {'easy', 'medium', 'hard'};

class PackQuestion {
  const PackQuestion({
    required this.id,
    required this.type,
    required this.prompt,
    required this.acceptedAnswers,
    required this.timeLimitMs,
    this.imageUrl,
    this.difficulty = 'easy',
  });

  final String id;
  final String type;
  final String prompt;
  final List<String> acceptedAnswers;
  final int timeLimitMs;

  /// On LAN this points at the hosting device itself, e.g.
  /// `http://192.168.1.20:4040/api/room-images/<key>` — reachable on the
  /// local network only, but an ordinary URL as far as any client is
  /// concerned (PROTOCOL.md §5.1).
  final String? imageUrl;
  final String difficulty;

  PackQuestion copyWith({int? timeLimitMs, String? id}) => PackQuestion(
    id: id ?? this.id,
    type: type,
    prompt: prompt,
    acceptedAnswers: acceptedAnswers,
    timeLimitMs: timeLimitMs ?? this.timeLimitMs,
    imageUrl: imageUrl,
    difficulty: difficulty,
  );
}

class Pack {
  const Pack({
    required this.titles,
    required this.questions,
    this.defaultTimeLimitMs,
    this.defaultDifficultyMultiplier = false,
  });

  const Pack.empty()
    : titles = const [],
      questions = const [],
      defaultTimeLimitMs = null,
      defaultDifficultyMultiplier = false;

  /// Builds a pack from a quiz document held on the device.
  ///
  /// Questions without an explicit id get a positional one, matching what the
  /// server does when it stores a quiz: ids only have to be stable within the
  /// room, and clients must not parse them (PROTOCOL.md §1).
  ///
  /// [imageUrl] turns a stored photo key into the URL clients fetch it from,
  /// the same seam `Fazoura.Quizzes.to_pack/2` has. Without it a photo
  /// question keeps its prompt and scores normally but carries no image —
  /// which is all a pack built outside a running host can honestly say.
  factory Pack.fromQuiz(
    QuizDocument quiz, {
    String Function(String key)? imageUrl,
  }) {
    final questions = quiz.questions ?? const <QuizQuestion>[];
    return Pack(
      titles: [quiz.title],
      defaultTimeLimitMs: quiz.defaultSettings.timeLimitMs,
      defaultDifficultyMultiplier: quiz.defaultSettings.difficultyMultiplier,
      questions: [
        for (final (index, question) in questions.indexed)
          PackQuestion(
            id: question.id ?? 'q${index + 1}',
            type: question.type,
            prompt: question.prompt,
            acceptedAnswers: question.acceptedAnswers,
            timeLimitMs: question.timeLimitMs ?? _defaultTimeLimitMs,
            imageUrl: _imageUrl(question, imageUrl),
            difficulty: _difficulty(question.difficulty),
          ),
      ],
    );
  }

  /// One pool from the quizzes the host selected, in the order they selected
  /// them — the counterpart of `Fazoura.Game.Pack.merge/1` (PROTOCOL.md §6.4).
  ///
  /// Question ids are rewritten to stay unique across the pool: two quizzes
  /// may each call a question `q1`, and clients use the id to tell one
  /// question from the next. Ids are opaque (§1), so rewriting them is ours to
  /// do. Lobby defaults come from the first quiz, because a pool has none.
  factory Pack.merge(List<Pack> packs) {
    if (packs.length == 1) return packs.single;
    return Pack(
      titles: [for (final pack in packs) ...pack.titles],
      defaultTimeLimitMs: packs.first.defaultTimeLimitMs,
      defaultDifficultyMultiplier: packs.first.defaultDifficultyMultiplier,
      questions: [
        for (final (index, pack) in packs.indexed)
          for (final question in pack.questions)
            question.copyWith(id: '$index-${question.id}'),
      ],
    );
  }

  /// Builds a pack straight from the JSON shape the shared contract fixtures
  /// use, the counterpart of `Fazoura.Game.Pack.from_map/1`. Only the fixture
  /// runner needs it: a real pack comes from a quiz document.
  factory Pack.fromMap(Map<String, dynamic> map) => Pack(
    titles: [map['title'] as String],
    questions: [
      for (final question in map['questions'] as List)
        _questionFromMap(question as Map<String, dynamic>),
    ],
  );

  static PackQuestion _questionFromMap(Map<String, dynamic> map) {
    final difficulty = map['difficulty'] as String? ?? 'easy';
    if (!_difficulties.contains(difficulty)) {
      throw ArgumentError.value(
        difficulty,
        'difficulty',
        'unknown difficulty; use easy, medium or hard',
      );
    }
    return PackQuestion(
      id: map['id'] as String,
      type: map['type'] as String? ?? 'text',
      prompt: map['prompt'] as String,
      acceptedAnswers: [
        for (final answer in map['accepted_answers'] as List) answer as String,
      ],
      timeLimitMs: map['time_limit_ms'] as int? ?? _defaultTimeLimitMs,
      imageUrl: map['image_url'] as String?,
      difficulty: difficulty,
    );
  }

  /// Titles of the selected quizzes, in the order the host chose them. Empty
  /// means nothing is selected yet (PROTOCOL.md §5.1, §6.4).
  final List<String> titles;
  final List<PackQuestion> questions;
  final int? defaultTimeLimitMs;
  final bool defaultDifficultyMultiplier;

  static String _difficulty(String value) =>
      _difficulties.contains(value) ? value : 'easy';

  /// Non-null only for a `text_photo` question whose photo was actually
  /// stored: `image_url` is what tells a client there is an image to show
  /// (PROTOCOL.md §5.1).
  static String? _imageUrl(
    QuizQuestion question,
    String Function(String key)? imageUrl,
  ) {
    final key = question.image?.key;
    if (!question.hasPhoto || key == null || imageUrl == null) return null;
    return imageUrl(key);
  }
}
